### PR TITLE
fix(liveness): add custom signer to client

### PR DESCRIPTION
**Pull request by calebpollman**
#### Description of changes
**Rev 1**
Add custom `Signer` to override hardcoded request expiry time in AWS JS SDK ([reference](https://github.com/aws/aws-sdk-js-v3/blob/93f81c4f1219a10a98cce2d71df42d5d04113ff1/packages/middleware-websocket/src/WebsocketSignatureV4.ts#L32)).

**Rev 2**
Consolidate client initialization logic in `createStreamingClient` and add unit testing:
- move client related utils in to _createStreamingClient_ dir
- cleanup `credentials` related types
- bump unit test thresholds

Add `resolveCredentials` utility to consolidate `credentials` retrieval logic:
- improve testability of error conditions in `credentials` retrieval

Pass `Signer` to `RekognitionStreamingClient` as `signerConstructor` instead of `signer` param:
- dedupe passing of `credentials` and `region` to `Signer` and `RekognitionStreamingClient`
- remove requirement for manual assignment of `sha256` and `service` params

**Rev3**
Refactor `resolveCredentials` to return valid `credentialsProvider` to ensure existing `credentials` resolving behavior is maintained
- remove no longer necessary `@aws-crypto/sha256-js` dep
- remove extraneous allowed bundle size increase

#### Description of how you validated changes
Manual verification of expiry time in request headers, manual test of UI component, add unit tests
#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [x] `yarn test` passes and tests are updated/added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/react-liveness/src/components/FaceLivenessDetector/service/utils/Signer.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  ): Promise<HttpRequest> {
+    return super.presign(request, {
+      ...options,
+      expiresIn: REQUEST_EXPIRY,
+      // `headers` that should not be signed. Transcribe-streaming WebSocket
+      // request omits `headers` except for required `host` header. Signature
+      // could be a mismatch if other `headers` are signed
+      unsignableHeaders: new Set(
+        Object.keys(request.headers).filter((header) => header !== 'host')
+      ),
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/service/utils/Signer.ts:37]: Copied from AWS JS SDK: https://github.com/aws/aws-sdk-js-v3/blob/93f81c4f1219a10a98cce2d71df42d5d04113ff1/packages/middleware-websocket/src/WebsocketSignatureV4.ts#L33-L36

**Review by AllanZhengYP — APPROVED** [review]
The change looks good to me. However, it change is better to be ported to AWS SDK. Here we are adding marginal risk by depending on packages not claimed to be supported publicly e.g. `Sha256`.

On a separate note, the credentials for the Rekognition client ([link](https://github.com/aws-amplify/amplify-ui/pull/5063/files#diff-cb25775a5ae955320cb75ef63794d5779468a5a02f9f1695534635c1f72b8c78R127)) seems to use static credentials instead of a credential provider. This might cause the credential to expire if the page is not refreshed for > 1 hour.

**Comment by calebpollman** [comment]
> The change looks good to me. However, it change is better to be ported to AWS SDK. Here we are adding marginal risk by depending on packages not claimed to be supported publicly e.g. `Sha256`.

Agree for sure. Will discuss with service team after verification that this fix works as expected

File: yarn.lock [hunk 2]
```diff
@@ -608,7 +608,7 @@
     "@aws-sdk/types" "^3.222.0"
     tslib "^1.11.1"
 
-"@aws-crypto/sha256-js@5.2.0":
+"@aws-crypto/sha256-js@5.2.0", "@aws-crypto/sha256-js@^5.2.0":
```
Comment (calebpollman) [inline yarn.lock:611]:
```suggestion
"@aws-crypto/sha256-js@5.2.0":

```

Comment (calebpollman) [inline yarn.lock:611]:
```suggestion
"@aws-crypto/sha256-js@5.2.0":

```

File: packages/react-liveness/src/components/FaceLivenessDetector/service/utils/streamProvider.ts [hunk 3]
```diff
@@ -114,40 +109,17 @@ export class LivenessStreamProvider {
...
-    const credentials =
-      this.credentialProvider ?? (await fetchAuthSession()).credentials;
-
-    if (!credentials) {
-      throw new Error('No credentials');
-    }
-
-    const clientconfig: RekognitionStreamingClientConfig = {
-      credentials,
+    this._client = await createStreamingClient({
```
Comment (esauerbo) [inline packages/react-liveness/src/components/FaceLivenessDetector/service/utils/streamProvider.ts:119]: Nice

**Review by esauerbo — DISMISSED** [review]

**Review by esauerbo — DISMISSED** [review]

File: packages/react-liveness/src/components/FaceLivenessDetector/service/utils/createStreamingClient/Signer.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export class Signer extends SignatureV4 {
+  public presign(
+    request: HttpRequest,
+    options?: Omit<RequestPresigningArguments, 'expiresIn'>
+  ): Promise<HttpRequest> {
+    return super.presign(request, {
+      ...options,
+      expiresIn: REQUEST_EXPIRY,
+      // `headers` that should not be signed. Transcribe-streaming WebSocket
```
Comment (AllanZhengYP) [inline packages/react-liveness/src/components/FaceLivenessDetector/service/utils/createStreamingClient/Signer.ts:18]: Very nit/ non-blocking:
```suggestion
      // `headers` that should not be signed. Liveness WebSocket
```

**Review by AllanZhengYP — APPROVED** [review]
Looks good to me now! Thanks for addressing all the concern in single PR!

File: packages/react-liveness/package.json [hunk 5]
```diff
@@ -80,7 +81,7 @@
       "name": "FaceLivenessDetector",
       "path": "dist/esm/index.mjs",
       "import": "{ FaceLivenessDetector }",
-      "limit": "275 kB"
+      "limit": "279 kB"
```
Comment (calebpollman) [inline packages/react-liveness/package.json:84]:
```suggestion
      "limit": "280 kB"
```

**Review by esauerbo — DISMISSED** [review]

File: packages/react-liveness/src/components/FaceLivenessDetector/service/utils/createStreamingClient/Signer.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  ): Promise<HttpRequest> {
+    return super.presign(request, {
+      ...options,
+      expiresIn: REQUEST_EXPIRY,
+      // `headers` that should not be signed. Liveness WebSocket
+      // request omits `headers` except for required `host` header. Signature
+      // could be a mismatch if other `headers` are signed
+      unsignableHeaders: new Set(
+        Object.keys(request.headers).filter((header) => header !== 'host')
+      ),
```
Comment (calebpollman) [inline packages/react-liveness/src/components/FaceLivenessDetector/service/utils/createStreamingClient/Signer.ts:23]: See https://github.com/aws-amplify/amplify-ui/pull/5063#discussion_r1520511593

**Review by esauerbo — APPROVED** [review]

**Review by reesscot — APPROVED** [review]

**Pull request merged**